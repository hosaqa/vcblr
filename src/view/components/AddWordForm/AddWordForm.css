.AddWordForm {
}

.AddWordForm__item {
  width: 100%;
}
