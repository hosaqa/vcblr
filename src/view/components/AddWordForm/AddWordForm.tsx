import React, { useState, useCallback } from 'react';
import { inject, observer } from 'mobx-react';
import Form from '../../UI/Form';
import Button from '../../UI/Button';
import TextField from '../../UI/TextField';
import './AddWordForm.css';
import { wordsStoreKey, IWordsStore } from '../../../stores/WordsStore';

interface IAddWordForm {
  wordsStore?: IWordsStore;
}

const AddWordForm: React.FC<IAddWordForm> = ({ wordsStore }) => {
  console.log(wordsStore);
  const { addWord, wordIsExists } = wordsStore!;

  const initialState = {
    wordEng: '',
    wordEngIsExists: false,
    wordRu: '',
    wordRuIsExists: false,
    contextEng: '',
    contextRu: '',
  };

  const [state, setState] = useState(initialState);

  const {
    wordEng,
    wordEngIsExists,
    wordRu,
    wordRuIsExists,
    contextEng,
    contextRu,
  } = state;

  const resetState = () => setState(initialState);

  const inputOnChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value, name } = event.target;

      const lang = event.target.getAttribute('data-lang')!;

      const existKey = lang === 'eng' ? 'wordEngIsExists' : 'wordRuIsExists';

      setState({
        ...state,
        [name]: value,
        [existKey]: wordIsExists(lang, value),
      });
    },
    [state]
  );

  const handleSubmit = useCallback(
    (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      addWord({
        eng: wordEng,
        ru: wordRu,
        contextRu,
        contextEng,
      });

      resetState();
    },
    [state]
  );

  const submitIsForbidden =
    !wordEng.length || !wordRu.length || wordEngIsExists || wordRuIsExists;

  return (
    <Form className="AddWordForm" onSubmit={handleSubmit}>
      <h1>Add your word</h1>
      <div className="Form__row">
        <label htmlFor="AddWordForm_eng">Word Eng</label>
        <TextField
          name="wordEng"
          className="AddWordForm__item"
          data-lang="eng"
          value={wordEng}
          onChange={inputOnChange}
          autoComplete="off"
        />
        {wordEngIsExists && (
          <small className="error-text">
            This word already exists in the database
          </small>
        )}
      </div>
      <div className="Form__row">
        <label htmlFor="AddWordForm_ru">Word Ru</label>
        <TextField
          name="wordRu"
          className="AddWordForm__item"
          data-lang="ru"
          value={wordRu}
          onChange={inputOnChange}
          autoComplete="off"
        />
        {wordRuIsExists && (
          <small className="error-text">
            This word already exists in the database
          </small>
        )}
      </div>
      <div className="Form__row">
        <label htmlFor="AddWordForm_eng-context">Context Eng</label>
        <TextField
          name="contextEng"
          className="AddWordForm__item"
          value={contextEng}
          onChange={inputOnChange}
          autoComplete="off"
        />
      </div>
      <div className="Form__row">
        <label htmlFor="AddWordForm_ru-context">Context Ru</label>
        <TextField
          name="contextRu"
          className="AddWordForm__item"
          value={contextRu}
          onChange={inputOnChange}
          autoComplete="off"
        />
      </div>
      <div className="Form__row">
        <Button
          disabled={submitIsForbidden}
          type="submit"
          className="AddWordForm__item"
        >
          Add word
        </Button>
      </div>
    </Form>
  );
};

export default inject(wordsStoreKey)(observer(AddWordForm));
