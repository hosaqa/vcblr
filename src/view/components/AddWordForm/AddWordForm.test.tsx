import React from 'react';
import { shallow } from 'enzyme';
import AddWordForm from './AddWordForm';

const WordsStore = {
  wordList: [
    {
      eng: 'stockpile',
      ru: 'запас оружия',
      contextRu: 'Это наш личный запас оружия на леднековый период',
      contextEng: 'This is our private stokpile for the Ice age.',
      dateAdded: '2020-02-12T21:20:16.688Z',
    },
  ],
  addWord: jest.fn(),
  wordIsExists: jest.fn(),
};

describe('AddWordForm', () => {
  const props = {
    wordsStore: WordsStore,
  };

  const initialState = {
    wordEng: '',
    wordEngIsExists: false,
    wordRu: '',
    wordRuIsExists: false,
    contextEng: '',
    contextRu: '',
  };

  describe('AddWordForm initial rendering', () => {
    const addWordForm = shallow(<AddWordForm {...props} />);

    it('renders properly', () => {
      expect(addWordForm).toMatchSnapshot();
    });
  });
});
