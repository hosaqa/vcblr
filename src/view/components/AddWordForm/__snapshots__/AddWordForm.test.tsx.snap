// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddWordForm AddWordForm initial rendering renders properly 1`] = `ShallowWrapper {}`;
